src/chainstate/stacks/index/cache.rs

```diff
@@ -0,0 +1,877 @@
+// Copyright (C) 2013-2020 Blockstack PBC, a public benefit corporation
+// Copyright (C) 2020-2022 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::char::from_digit;
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::convert::{TryFrom, TryInto};
+use std::env;
+use std::fmt;
+use std::fs;
+use std::hash::{Hash, Hasher};
+use std::io;
+use std::io::{BufWriter, Cursor, Read, Seek, SeekFrom, Write};
+use std::iter::FromIterator;
+use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut};
+use std::os;
+use std::path::{Path, PathBuf};
+use std::time::SystemTime;
+use std::{cmp, error};
+
+use rusqlite::{
+    types::{FromSql, ToSql},
+    Connection, Error as SqliteError, ErrorCode as SqliteErrorCode, OpenFlags, OptionalExtension,
+    Transaction, NO_PARAMS,
+};
+
+use chainstate::stacks::index::bits::{
+    get_node_byte_len, get_node_hash, read_block_identifier, read_hash_bytes, read_node_hash_bytes,
+    read_nodetype, read_root_hash, write_nodetype_bytes,
+};
+use chainstate::stacks::index::node::{
+    clear_backptr, is_backptr, set_backptr, TrieNode, TrieNode16, TrieNode256, TrieNode4,
+    TrieNode48, TrieNodeID, TrieNodeType, TriePath, TriePtr,
+};
+use chainstate::stacks::index::Error;
+use chainstate::stacks::index::TrieLeaf;
+use chainstate::stacks::index::{trie_sql, ClarityMarfTrieId, MarfTrieId};
+use util_lib::db::sql_pragma;
+use util_lib::db::sqlite_open;
+use util_lib::db::tx_begin_immediate;
+use util_lib::db::tx_busy_handler;
+use util_lib::db::Error as db_error;
+use util_lib::db::SQLITE_MMAP_SIZE;
+
+use stacks_common::types::chainstate::BlockHeaderHash;
+use stacks_common::types::chainstate::BLOCK_HEADER_HASH_ENCODED_SIZE;
+use stacks_common::types::chainstate::{TrieHash, TRIEHASH_ENCODED_SIZE};
+
+/// Fully-qualified address of a Trie node.  Includes both the block ID and the pointer within the
+/// block's blob as to where it is stored.
```

Comment:
The Block ID is the `u32`? There's a concept of a `MarfTrieId` in the codebase, but that's a `[u8; 32]`, so what is the `block ID`?

Reply:
Sorry, that's the block's local ID -- i.e. the row ID in the database.  Will update the comment.